lib/os_io.c

```diff
@@ -37,6 +37,197 @@ ___io_module ___io_mod =
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = size;
+
+  return 1;
+
+ error:
+  if (readfds != NULL)
+    ___FREE_MEM (readfds);
+  if (writefds != NULL)
+    ___FREE_MEM (writefds);
+  if (exceptfds != NULL)
+    ___FREE_MEM (exceptfds);
+
+  return 0;
+}
+
+___HIDDEN int ___fdset_realloc
+   ___P((___processor_state ps,
+         int newsize),
+        (ps,
+         newsize)
+        ____processor_state ps;
+        int newsize;)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  void *exceptfds = NULL;
+  int oldsize = ps->os.fdset.size;
+  int oldbytes;
+  int newbytes;
+
+  if (newsize <= oldsize) /* we don't shrink */
+    return 1;
+
+  oldbytes = ___CEILING_DIV (oldsize,8);
+  newbytes = ___CEILING_DIV (newsize,8);
+
+  readfds  = ___ALLOC_MEM (newbytes);
+  if (readfds == NULL)
+    goto error;
+
+  writefds = ___ALLOC_MEM (newbytes);
+  if (writefds == NULL)
+    goto error;
+
+#ifdef USE_select
+  exceptfds = ___ALLOC_MEM (newbytes);
+  if (exceptfds == NULL)
+    goto error;
+#endif
+
+  memcpy (readfds, ps->os.fdset.readfds, oldbytes);
+  memset (___CAST(char*, readfds) + oldbytes, 0, newbytes - oldbytes);
+  memcpy (writefds, ps->os.fdset.writefds, oldbytes);
+  memset (___CAST(char*, writefds) + oldbytes, 0, newbytes - oldbytes);
+#ifdef USE_select
+  memcpy (exceptfds, ps->os.fdset.exceptfds, oldbytes);
+  memset (___CAST(char*, exceptfds) + oldbytes, 0, newbytes - oldbytes);
+#endif
+
+  ___FREE_MEM (ps->os.fdset.readfds);
+  ___FREE_MEM (ps->os.fdset.writefds);
+#ifdef USE_select
+  ___FREE_MEM (ps->os.fdset.exceptfds);
+#endif
+
+  ps->os.fdset.readfds = readfds;
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = newsize;
+
+  return 1;
+
+ error:
+  if (readfds != NULL)
+    ___FREE_MEM (readfds);
+  if (writefds != NULL)
+    ___FREE_MEM (writefds);
+  if (exceptfds != NULL)
+    ___FREE_MEM (exceptfds);
+
+  return 0;
+}
+
+___HIDDEN int ___fdset_size
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ps->os.fdset.size;
+}
+
+___HIDDEN ___fdbits *___fdset_readfds
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ___CAST(___fdbits*, ps->os.fdset.readfds);
+}
+
+___HIDDEN ___fdbits *___fdset_writefds
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ___CAST(___fdbits*, ps->os.fdset.writefds);
+}
+
+___HIDDEN ___fdbits *___fdset_exceptfds
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ___CAST(___fdbits*, ps->os.fdset.exceptfds);
+}
+
+#endif
+
+#ifdef USE_POSIX
+
+void ___fdset_clear_overflow_pstate
+   ___P((___processor_state ___ps),
+        (___ps)
+___processor_state ___ps;)
+{
+#ifdef USE_select_or_poll
+
+  ___virtual_machine_state ___vms = ___VMSTATE_FROM_PSTATE(___ps);
+
+  if (___PROCESSOR_ID(___ps, ___vms) == 0)
```

Comment:
Indeed the check for processor 0 is elsewhere.  Moreover the parameter of this function should be ```___virtual_machine_state ___vms``` and the function should be named ```___fdset_clear_overflow_vmstate```.